### DISCUSSION TITLE
Plugin Works about 80% with Air Conditioner

### DISCUSSION THREAD

**Discussion by plymouthvan** [nVuln/homebridge-lg-thinq#93] [General]
Love this plugin! Being able to automatically turn the air conditioner off when I leave the house is saving me a bunch of money on my energy bill. It works about 80% for my LW1217ERSM air conditioner.

A few things don't work though, or at least not reliably:

- Switching between auto mode and cooling mode is hit or miss, mostly miss. It will sometimes allow switching to cooling mode from auto mode, but never the other way around from cooling mode to auto mode. The main thing I need this plugin for is to turn the unit on and off, but I do switch between modes a lot and it would be nice to be able to do it from the phone.

- Also, the exact temperature setting is fickle. For example setting the unit to 73 degrees in Homekit will set it to 74 on the unit. Setting it to 68 degrees in HomeKit might set the unit to 67 or 69, but it never lands right on what it was set to in HomeKit.

Is there anything I can do to fix this behavior?

**Comment by nVuln** [comment]
> Switching between auto mode and cooling mode is hit or miss, mostly miss. It will sometimes allow switching to cooling mode from auto mode, but never the other way around from cooling mode to auto mode. The main thing I need this plugin for is to turn the unit on and off, but I do switch between modes a lot and it would be nice to be able to do it from the phone.

you can enable `ac_fan_control = true`, then you can turn on/off auto mode inside this fan service (the button near swing button)

> Also, the exact temperature setting is fickle. For example setting the unit to 73 degrees in Homekit will set it to 74 on the unit. Setting it to 68 degrees in HomeKit might set the unit to 67 or 69, but it never lands right on what it was set to in HomeKit.

a reason is Homekit handled temperature as celcius only, even you selected farenheit on setting
for example, when you select 74 F on Home app, homekit will convert it to C and send this value to Homebridge (this plugin), the plugin will receive celcius value and converted it to farenheit and send it to unit => may be something wrong between this

**Reply by plymouthvan** [reply]
Thank you for the explanation on the Celsius conversion, that makes sense. It's not really an issue, just an oddity.

Although I can live without it, auto mode is a more significant nuisance. AC_FAN_CONTROL is actually already set to TRUE, and that's where the wonkiness starts. As long as that setting is disabled, the basic on/off and temperature setting works near flawlessly. But that changes when the setting is enabled. I can see auto mode in the accessories screen as a separate device, and although it does occasionally turn auto mode off, it never turns it on. Or maybe it's the other way around; regardless, under any conditions, it only does anything about 1/20 times, so it's hard to get a fix on exactly what it is and isn't doing. Enabling fan control also does some weird stuff, like even though AC_MODE is set to COOLING, when fan control is enabled, HomeKit seems to think it's both a heater and a cooler, and sometimes reports that it's trying to heat the room. I was figuring this all might just have to do with a unit that hasn't been tested, but I'm pretty new to home bridge stuff.

Any insight on how I might trouble shoot this? Or better to just let it go and hope for the best in some future update?

**Reply by nVuln** [reply]
> like even though AC_MODE is set to COOLING, when fan control is enabled, HomeKit seems to think it's both a heater and a cooler, and sometimes reports that it's trying to heat the room.

is your AC working in both mode COOL and HEAT ?
please turn on homebridge debug mode, then upload debug logging after you make some controlling
also upload some screenshot from Home app

**Reply by plymouthvan** [reply]
I'm not entirely sure what you're asking--the Air Conditioner Unit itself only supports cooling. The setting in the Homebridge plugin is set to "cooling" only. Here's a screenshot from the HomeKit app, you can see it says "heating" at the top left.

![Screen Shot 2021-10-26 at 3 30 03 PM](https://user-images.githubusercontent.com/22523350/138948752-3c489de3-3bd4-4abf-afd6-2c4ddf933a4f.png)

I'm not 100% sure this is what you're asking for, but here is the red text the home bridge console is spitting out:

[26/10/2021, 15:32:08] [homebridge-lg-thinq] Set fan speed =  5
[26/10/2021, 15:32:08] [homebridge-lg-thinq] Set fan speed =  5
[26/10/2021, 15:32:12] [homebridge-lg-thinq] Set fan speed =  2
[26/10/2021, 15:32:12] [homebridge-lg-thinq] Set fan speed =  1
[26/10/2021, 15:32:12] [homebridge-lg-thinq] Set fan speed =  1
[26/10/2021, 15:32:59] [homebridge-lg-thinq] Unknown Error:  {
  status: 400,
  statusText: '400',
  headers: {
    date: 'Tue, 26 Oct 2021 19:32:59 GMT',
    'content-type': 'application/json;charset=UTF-8',
    'content-length': '42',
    connection: 'close',
    'x-dynatrace-js-agent': 'true',
    'set-cookie': [
      'dtCookie=31FFBA4C049557669DDA97D08ECDB6EB;domain=lgthinq.com;path=/',
      'dtCookie=31FFBA4C049557669DDA97D08ECDB6EB; Path=/; Domain=.lgthinq.com'
    ]
  },
  config: {
    url: 'https://aic-service.lgthinq.com:46030/v1/service/devices/60af636e-da89-1ef1-a1ff-1c3929046b29/control-sync',
    method: 'post',
    data: '{"ctrlKey":"basicCtrl","command":"Set","dataKey":"airState.opMode","dataValue":6}',
    headers: {
      Accept: 'application/json, text/plain, */*',
      'Content-Type': 'application/json',
      'x-api-key': 'VGhpblEyLjAgU0VSVklDRQ==',
      'x-thinq-app-ver': '3.6.1200',
      'x-thinq-app-type': 'NUTS',
      'x-thinq-app-level': 'PRD',
      'x-thinq-app-os': 'ANDROID',
      'x-thinq-app-logintype': 'LGE',
      'x-service-code': 'SVC202',
      'x-country-code': 'US',
      'x-language-code': 'en-US',
      'x-service-phase': 'OP',
      'x-origin': 'app-native',
      'x-model-name': 'samsung/SM-G930L',
      'x-os-version': 'AOS/7.1.2',
      'x-app-version': 'LG ThinQ/3.6.12110',
      'x-message-id': 'arpHCS5An4WQJSGHQfcCus',
      'user-agent': 'okhttp/3.14.9',
      'x-emp-token': '5807000920a1d01de57914248834469867b3d89e4c708736db81d1aea121cf02981bb7b076e2668d406e8a388a9d2d66',
      'x-user-no': 'US2105219331539',
      'x-client-id': 'cd76fdb4f4d680e05d00fcd1f0d17fbdbb1b137b467b2384c77f728e3729992e',
      'Content-Length': 81
    },
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 60000,
    adapter: [Function: httpAdapter],
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    validateStatus: [Function: validateStatus],
    transitional: {
      silentJSONParsing: true,
      forcedJSONParsing: true,
      clarifyTimeoutError: false
    },
    'axios-retry': { retryCount: 0, lastRequestTime: 1635276758682 }
  },
  request: <ref *1> ClientRequest {
    _events: [Object: null prototype] {
      abort: [Function (anonymous)],
      aborted: [Function (anonymous)],
      connect: [Function (anonymous)],
      error: [Function (anonymous)],
      socket: [Function (anonymous)],
      timeout: [Function (anonymous)],
      prefinish: [Function: requestOnPrefinish]
    },
    _eventsCount: 7,
    _maxListeners: undefined,
    outputData: [],
    outputSize: 0,
    writable: true,
    destroyed: false,
    _last: true,
    chunkedEncoding: false,
    shouldKeepAlive: false,
    _defaultKeepAlive: true,
    useChunkedEncodingByDefault: true,
    sendDate: false,
    _removedConnection: false,
    _removedContLen: false,
    _removedTE: false,
    _contentLength: null,
    _hasBody: true,
    _trailer: '',
    finished: true,
    _headerSent: true,
    socket: TLSSocket {
      _tlsOptions: [Object],
      _secureEstablished: true,
      _securePending: false,
      _newSessionPending: false,
      _controlReleased: true,
      secureConnecting: false,
      _SNICallback: null,
      servername: 'aic-service.lgthinq.com',
      alpnProtocol: false,
      authorized: true,
      authorizationError: null,
      encrypted: true,
      _events: [Object: null prototype],
      _eventsCount: 10,
      connecting: false,
      _hadError: false,
      _parent: null,
      _host: 'aic-service.lgthinq.com',
      _readableState: [ReadableState],
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: false,
      _sockname: null,
      _pendingData: null,
      _pendingEncoding: '',
      server: undefined,
      _server: null,
      ssl: [TLSWrap],
      _requestCert: true,
      _rejectUnauthorized: true,
      parser: null,
      _httpMessage: [Circular *1],
      timeout: 60000,
      [Symbol(res)]: [TLSWrap],
      [Symbol(verified)]: true,
      [Symbol(pendingSession)]: null,
      [Symbol(async_id_symbol)]: 1120,
      [Symbol(kHandle)]: [TLSWrap],
      [Symbol(kSetNoDelay)]: false,
      [Symbol(lastWriteQueueSize)]: 0,
      [Symbol(timeout)]: Timeout {
        _idleTimeout: 60000,
        _idlePrev: [TimersList],
        _idleNext: [Timeout],
        _idleStart: 107532,
        _onTimeout: [Function: bound ],
        _timerArgs: undefined,
        _repeat: null,
        _destroyed: false,
        [Symbol(refed)]: false,
        [Symbol(kHasPrimitive)]: false,
        [Symbol(asyncId)]: 1128,
        [Symbol(triggerId)]: 1123
      },
      [Symbol(kBuffer)]: null,
      [Symbol(kBufferCb)]: null,
      [Symbol(kBufferGen)]: null,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesRead)]: 0,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(connect-options)]: [Object],
      [Symbol(RequestTimeout)]: undefined
    },
    _header: 'POST /v1/service/devices/60af636e-da89-1ef1-a1ff-1c3929046b29/control-sync HTTP/1.1\r\n' +
      'Accept: application/json, text/plain, */*\r\n' +
      'Content-Type: application/json\r\n' +
      'x-api-key: VGhpblEyLjAgU0VSVklDRQ==\r\n' +
      'x-thinq-app-ver: 3.6.1200\r\n' +
      'x-thinq-app-type: NUTS\r\n' +
      'x-thinq-app-level: PRD\r\n' +
      'x-thinq-app-os: ANDROID\r\n' +
      'x-thinq-app-logintype: LGE\r\n' +
      'x-service-code: SVC202\r\n' +
      'x-country-code: US\r\n' +
      'x-language-code: en-US\r\n' +
      'x-service-phase: OP\r\n' +
      'x-origin: app-native\r\n' +
      'x-model-name: samsung/SM-G930L\r\n' +
      'x-os-version: AOS/7.1.2\r\n' +
      'x-app-version: LG ThinQ/3.6.12110\r\n' +
      'x-message-id: arpHCS5An4WQJSGHQfcCus\r\n' +
      'user-agent: okhttp/3.14.9\r\n' +
      'x-emp-token: 5807000920a1d01de57914248834469867b3d89e4c708736db81d1aea121cf02981bb7b076e2668d406e8a388a9d2d66\r\n' +
      'x-user-no: US2105219331539\r\n' +
      'x-client-id: cd76fdb4f4d680e05d00fcd1f0d17fbdbb1b137b467b2384c77f728e3729992e\r\n' +
      'Content-Length: 81\r\n' +
      'Host: aic-service.lgthinq.com:46030\r\n' +
      'Connection: close\r\n' +
      '\r\n',
    _keepAliveTimeout: 0,
    _onPendingData: [Function: noopPendingOutput],
    agent: Agent {
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      defaultPort: 443,
      protocol: 'https:',
      options: [Object],
      requests: {},
      sockets: [Object],
      freeSockets: {},
      keepAliveMsecs: 1000,
      keepAlive: false,
      maxSockets: Infinity,
      maxFreeSockets: 256,
      scheduling: 'lifo',
      maxTotalSockets: Infinity,
      totalSocketCount: 1,
      maxCachedSessions: 100,
      _sessionCache: [Object],
      [Symbol(kCapture)]: false
    },
    socketPath: undefined,
    method: 'POST',
    maxHeaderSize: undefined,
    insecureHTTPParser: undefined,
    path: '/v1/service/devices/60af636e-da89-1ef1-a1ff-1c3929046b29/control-sync',
    _ended: true,
    res: IncomingMessage {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 3,
      _maxListeners: undefined,
      socket: [TLSSocket],
      httpVersionMajor: 1,
      httpVersionMinor: 1,
      httpVersion: '1.1',
      complete: true,
      headers: [Object],
      rawHeaders: [Array],
      trailers: {},
      rawTrailers: [],
      aborted: false,
      upgrade: false,
      url: '',
      method: null,
      statusCode: 400,
      statusMessage: '400',
      client: [TLSSocket],
      _consuming: false,
      _dumped: false,
      req: [Circular *1],
      responseUrl: 'https://aic-service.lgthinq.com:46030/v1/service/devices/60af636e-da89-1ef1-a1ff-1c3929046b29/control-sync',
      redirects: [],
      [Symbol(kCapture)]: false,
      [Symbol(RequestTimeout)]: undefined
    },
    aborted: false,
    timeoutCb: null,
    upgradeOrConnect: false,
    parser: null,
    maxHeadersCount: null,
    reusedSocket: false,
    host: 'aic-service.lgthinq.com',
    protocol: 'https:',
    _redirectable: Writable {
      _writableState: [WritableState],
      _events: [Object: null prototype],
      _eventsCount: 3,
      _maxListeners: undefined,
      _options: [Object],
      _ended: true,
      _ending: true,
      _redirectCount: 0,
      _redirects: [],
      _requestBodyLength: 81,
      _requestBodyBuffers: [],
      _onNativeResponse: [Function (anonymous)],
      _currentRequest: [Circular *1],
      _currentUrl: 'https://aic-service.lgthinq.com:46030/v1/service/devices/60af636e-da89-1ef1-a1ff-1c3929046b29/control-sync',
      _timeout: null,
      [Symbol(kCapture)]: false
    },
    [Symbol(kCapture)]: false,
    [Symbol(kNeedDrain)]: false,
    [Symbol(corked)]: 0,
    [Symbol(kOutHeaders)]: [Object: null prototype] {
      accept: [Array],
      'content-type': [Array],
      'x-api-key': [Array],
      'x-thinq-app-ver': [Array],
      'x-thinq-app-type': [Array],
      'x-thinq-app-level': [Array],
      'x-thinq-app-os': [Array],
      'x-thinq-app-logintype': [Array],
      'x-service-code': [Array],
      'x-country-code': [Array],
      'x-language-code': [Array],
      'x-service-phase': [Array],
      'x-origin': [Array],
      'x-model-name': [Array],
      'x-os-version': [Array],
      'x-app-version': [Array],
      'x-message-id': [Array],
      'user-agent': [Array],
      'x-emp-token': [Array],
      'x-user-no': [Array],
      'x-client-id': [Array],
      'content-length': [Array],
      host: [Array]
    }
  },
  data: { resultCode: '9006', result: { data: '' } }
}
[26/10/2021, 15:32:59] [homebridge-lg-thinq] This plugin generated a warning from the characteristic 'Target Heater-Cooler State': characteristic value 1 is not contained in valid values array. See https://git.io/JtMGR for more info.
[26/10/2021, 15:32:59] [homebridge-lg-thinq] This plugin generated a warning from the characteristic 'Target Heater-Cooler State': characteristic value 1 is not contained in valid values array. See https://git.io/JtMGR for more info.
